imap/jmap_ical.c

```diff
@@ -110,11 +110,125 @@ static int is_valid_jmapid(const char *s)
     return i > 0 && s[i] == '\0';
 }
 
+/* A helper structure to organize iCalendar components */
+struct icalcomps {
+    hash_table by_uidrecurid; /* pointer to icalcomponent */
+    hash_table by_uid; /* ptrarray_t of icalcomponent */
+    struct buf buf;
+};
+
+#define ICALCOMPS_INITIALIZER { \
+    HASH_TABLE_INITIALIZER, \
+    HASH_TABLE_INITIALIZER, \
+    BUF_INITIALIZER \
+}
+
+static const char *make_uidrecurid(icalcomponent *comp, struct buf *buf)
+{
+    const char *uid = icalcomponent_get_uid(comp);
+    if (!uid) return NULL;
+
+    icalproperty *recurid = icalcomponent_get_first_property(comp,
+            ICAL_RECURRENCEID_PROPERTY);
+    if (!recurid) return NULL;
+
+    buf_setcstr(buf, uid);
+    buf_putc(buf, ';');
+    // append complete prop, including any TZID
+    buf_appendcstr(buf, icalproperty_as_ical_string(recurid));
+    return buf_cstring(buf);
+}
+
+static void icalcomps_init(struct icalcomps *comps, icalcomponent *ical)
+{
+    // hash table requires us to preallocate its size
+    size_t ncomps = 0;
+    icalcomponent *comp;
+    for (comp = icalcomponent_get_first_component(ical, ICAL_VEVENT_COMPONENT);
+         comp;
+         comp = icalcomponent_get_next_component(ical, ICAL_VEVENT_COMPONENT)) {
+        ncomps++;
+    }
```

Comment:
This loop can probably be replaced with icalcomponent_count_components()

Reply:
great, done